tools/parity_scale_cli/src/main.rs

```diff
@@ -56,44 +60,81 @@ use iroha_primitives::{
     conststr::ConstString,
     unique_vec::UniqueVec,
 };
-use parity_scale_codec::DecodeAll;
+use parity_scale_codec::{DecodeAll, Encode};
+use serde::{de::DeserializeOwned, Serialize};
 
-/// Generate map with types and `dump_decoded()` ptr
-pub fn generate_map() -> DumpDecodedMap {
-    let mut map = DumpDecodedMap::new();
+/// Generate map with types and converter trait object
+fn generate_map() -> ConverterMap {
+    let mut map = ConverterMap::new();
 
     macro_rules! insert_into_map {
         ($t:ty) => {{
             let type_id = <$t as iroha_schema::TypeId>::id();
-
-            #[allow(trivial_casts)]
-            map.insert(type_id, <$t as DumpDecoded>::dump_decoded as DumpDecodedPtr)
+            map.insert(type_id, ConverterImpl::<$t>::new())
         }};
     }
 
     iroha_schema_gen::map_all_schema_types!(insert_into_map);
 
-    #[allow(trivial_casts)]
     map.insert(
         <iroha_schema::Compact<u128> as iroha_schema::TypeId>::id(),
-        <parity_scale_codec::Compact<u32> as DumpDecoded>::dump_decoded as DumpDecodedPtr,
+        ConverterImpl::<u32>::new(),
     );
 
     map
 }
 
+type ConverterMap = BTreeMap<String, Box<dyn Converter>>;
+
+struct ConverterImpl<T>(PhantomData<T>);
+
+impl<T> ConverterImpl<T> {
+    fn new() -> Box<Self> {
+        Box::new(Self(PhantomData))
+    }
+}
+
+trait Converter {
+    fn scale_to_rust(&self, input: &[u8]) -> Result<String>;
+    fn scale_to_json(&self, input: &[u8]) -> Result<String>;
+    fn json_to_scale(&self, input: &str) -> Result<Vec<u8>>;
+}
```

Comment:
Hm, the pattern of a map of types with dynamic `Conveter`s for each type seems to be an overkill for me, because it all goes down to a single `ConverterImpl` anyway.

I think overall implementation would be simpler if we have just a few generic functions:

```rust
mod convert {
	fn scale_to_rust<T: DecodeAll + Debug>(value: T, input: &[u8]) -> Result<String> { todo! }
	
	fn scale_to_json<T: DecodeAll + Serialize>(value: T, input: &[u8]) -> Result<String> { todo! }
	
	fn json_to_scale<T: DeserializeOwned + Encode>(value: T, input: &str) -> Result<Vec<u8>> { todo! }
}
```

(and also to replace `String`, `String`, and `Vec<u8>` with some newtypes like `RustDebug`, `Json`, and `Scale` to reduce blindness a bit)

Is there an issue I cannot recognise? Why did you chose the way that you chose?

Reply:
Just to say that we don't need `to_rust`. `to_json` is sufficient

Reply:
@0x009922 i'm not sure how you gonna convert string name of type the function you want to call. 
It could be some big `match` statement, but map in this case seems more readable to me.

Since our converter type doesn't have any state instead of `dyn` object we can just use function pointers.

```rust
struct Converter {
    scale_to_rust: fn(input: &[u8]) -> RustDebug,
    scale_to_json: fn(input: &[u8]) -> Json,
    json_to_scale: fn(input: &[u8]) -> Scale,
}
```

But since this program is essentially one off it doesn't really matter and i'm fine with `dyn` object approach.

Reply:
Aah, that is the part I missed, thanks.

Than I am fine too.